### ISSUE TITLE
Pytype seems do not check builtin map function

### ISSUE THREAD

**Issue by vuryleo** [google/pytype#449]
e.g.
```python
# map.py
def func(a: int) -> float:
    return float(a)

map(func, ['str'])
```

```bash
$ pytype map.py
Computing dependencies
Analyzing 1 sources with 0 local dependencies
ninja: Entering directory `/[redacted]/.pytype'
ninja: no work to do.
Success: no errors found
```

while
```bash
$ mypy map.py
map.py:5: error: Argument 1 to "map" has incompatible type "Callable[[int], float]"; expected "Callable[[str], float]"
Found 1 error in 1 file (checked 1 source file)
```

**Comment by rchen152** [comment]
Thanks for the report! Looks like the problem is that the `map()` definition here doesn't specify parameters for `function`:
https://github.com/google/pytype/blob/459ea901e0118cfbad5a7808622a4cc240498ac2/pytype/pytd/builtins/3/__builtin__.pytd#L79

**Comment by ghostoverflow** [comment]
@rchen152 Can I work on this?

**Comment by rchen152** [comment]
Of course, we'd love the help! This issue entails fixing the `map` definitions here:
https://github.com/google/pytype/blob/4109225011ddde177ee32af7da5832babaa52220/pytype/stubs/builtins/3/builtins.pytd#L89-L90
The typeshed definitions may be a helpful reference: https://github.com/python/typeshed/blob/b03cd49a8756dfd8e087dfb6b886225cc0afc950/stdlib/builtins.pyi#L994-L1028

Let me know if you run into any problems! One thing that has bitten other contributors before is that there are multiple copies of builtins.pytd (`builtins/2/builtins.pytd`, `builtins/2/__builtin__.pytd`, `builtins/3/builtins.pytd`, and `builtins/3/__builtin__.pytd` - phew!), so if you make a change and it doesn't seem to be reflected in tests, make sure you've edited all the builtins files for the version(s) the test is running in.

**Comment by ghostoverflow** [comment]
Thanks @rchen152 . I'm looking into it right now. Don you want me to replace these map definitions https://github.com/google/pytype/blob/4109225011ddde177ee32af7da5832babaa52220/pytype/stubs/builtins/3/builtins.pytd#L89-L90
With these ones?
https://github.com/python/typeshed/blob/b03cd49a8756dfd8e087dfb6b886225cc0afc950/stdlib/builtins.pyi#L994-L1028

I'll be thankful for your help.

**Comment by rchen152** [comment]
You might have to do some tweaking, but yeah, I'd start by replacing `def map(function: Callable[..., _T], *sequences: Iterable) -> Iterator[_T]: ...` with the series of overloads from the typeshed stub and see if that gets you the right behavior (of type-checking the Callable argument types).

**Comment by Jrryy** [comment]
I'm currently trying to work on this, but I've come across a problem I believe I should ask about.

Basically, I've modified the stub files according to python's own stubs so that maps would be correctly supported, and I've created a test case:
```python
# builtins.py/__builtins__.py
...
def map(__func: Callable[[_T], _S], __iter1: Iterable[_T]) -> Iterator[_S]: ...
...
```
```python
# Test case
from typing import Union
def func(a: Union[int, float, bool]) -> str:
    return str(a)
    
map(func, [1, 3.14, True])
```

So far so good, the tests pass correctly. However, if I add strings to all of this:

```python
from typing import Union
def func(a: Union[int, float, str, bool]) -> str:
    return str(a)
    
map(func, [1, 3.14, True, 'pi'])
```
The following error appears:
```
Line 5, in <module>: Built-in function map was called with the wrong arguments [wrong-arg-types]
         Expected: (__func, __iter1: Iterable[Union[bool, float, int, str]])
  Actually passed: (__func, __iter1: List[Union[bool, float, int, str]])
```
If I'm not wrong, the test case should be all right and the type checks should run without errors. `List` is also defined as a subtype of `Iterable` (`Iterable -> Sequence -> MutableSequence -> List`), so an argument with type `List` should be valid. Does this have something to do with [this section in the FAQ](https://github.com/google/pytype/blob/master/docs/faq.md#why-doesnt-str-match-against-string-iterables) or am I misunderstanding the error?

I've tested both examples I've presented with mypy too and something similar happens: the first one passes correctly but the second one outputs an error (`Argument 1 to "map" has incompatible type "Callable[[Union[int, float, str, bool]], str]"; expected "Callable[[object], str]"`), which is different but also appears when I use the type `str`. I feel like I'm not using `str` correctly but I don't really know why.

**Comment by rchen152** [comment]
Hmm, I think you've run into a bug in pytype. I cooked up an example unrelated to `map` that shows similar behavior, in which I think pytype is pretty clearly wrong:

```
from typing import Callable, Iterable, TypeVar, Union
T = TypeVar('T')
def f(x: Callable[[T], int], y: Iterable[T]):
  pass

def g(x: Union[int, str]):
  return 0

f(g, [0, ''])  # pytype reports [wrong-arg-types] here!
```

It's doing something funny when matching Unions to TypeVars, I bet. I've opened an issue (https://github.com/google/pytype/issues/1004) and will look into it some more.

Thanks for taking a stab at fixing this bug, and sorry you've hit this obstacle!
